File: crates/ans-protocol/src/name.rs
use sha2::{Digest, Sha256};

use crate::error::AnsError;

pub const ARCH_SUFFIX: &str = ".arch";
pub const NAME_HASH_DOMAIN: &[u8] = b"arch-name-service:name-hash:v1\0";

pub fn canonicalize_name(input: &str) -> Result<String, AnsError> {
    let label = input
        .strip_suffix(ARCH_SUFFIX)
        .ok_or(AnsError::InvalidSuffix)?;
    validate_label(label)?;
    Ok(format!("{label}{ARCH_SUFFIX}"))
}

pub fn validate_label(label: &str) -> Result<(), AnsError> {
    if !(1..=63).contains(&label.len()) {
        return Err(AnsError::InvalidLabelLength);
    }
    if label.starts_with('-') || label.ends_with('-') || label.contains("--") {
        return Err(AnsError::InvalidHyphenPlacement);
    }
    if !label
        .bytes()
        .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
    {
        return Err(AnsError::InvalidLabelCharacter);
    }
    Ok(())
}

pub fn name_hash(canonical_name: &str) -> Result<[u8; 32], AnsError> {
    let canonical_name = canonicalize_name(canonical_name)?;
    let mut hasher = Sha256::new();
    hasher.update(NAME_HASH_DOMAIN);
    hasher.update(canonical_name.as_bytes());
    Ok(hasher.finalize().into())
}

File: packages/ans-sdk/src/name.ts
import { ARCH_SUFFIX } from "./constants.js";
import { AnsError } from "./errors.js";
import { hashName } from "./hash.js";

export function validateLabel(label: string): void {
  if (label.length < 1 || label.length > 63) {
    throw new AnsError("InvalidLabelLength");
  }
  if (label.startsWith("-") || label.endsWith("-") || label.includes("--")) {
    throw new AnsError("InvalidHyphenPlacement");
  }
  for (const byte of new TextEncoder().encode(label)) {
    const isLower = byte >= 0x61 && byte <= 0x7a;
    const isDigit = byte >= 0x30 && byte <= 0x39;
    if (!isLower && !isDigit && byte !== 0x2d) {
      throw new AnsError("InvalidLabelCharacter");
    }
  }
}

/**
 * Normalize a `.arch` name to the single form that can exist on chain.
 *
 * The namespace is lowercase-only, like DNS: `Adams.arch` and `adams.arch` are
 * the same name, and only `adams.arch` is storable. Case folding therefore
 * belongs here, not at each call site — resolution used to reject any input a
 * user had capitalized. `validateLabel` stays strict because it checks a label
 * as stored, and the program must keep rejecting non-canonical input.
 */
export function canonicalizeName(input: string): string {
  const normalized = input.trim().toLowerCase();
  if (!normalized.endsWith(ARCH_SUFFIX)) {
    throw new AnsError("InvalidSuffix");
  }
  const label = normalized.slice(0, -ARCH_SUFFIX.length);
  validateLabel(label);
  return `${label}${ARCH_SUFFIX}`;
}

/** The canonical label (no suffix) for any accepted spelling of a name. */
export function canonicalizeLabel(label: string): string {
  return canonicalizeName(`${label.trim()}${ARCH_SUFFIX}`).slice(0, -ARCH_SUFFIX.length);
}

export function nameHash(input: string): Uint8Array {
  return hashName(canonicalizeName(input));
}
